src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,258 @@
+                return;
+
+            foreach (var previousResult in previousResults)
+            {
+                var textDocument = previousResult.TextDocument;
+                if (textDocument != null)
+                {
+                    var document = _solutionProvider.GetDocument(textDocument);
+                    if (document == null)
+                    {
+                        // Client is asking server about a document that no longer exists (i.e. was removed/deleted from the
+                        // workspace).  In that case we need to return an actual diagnostic report with `null` for the
+                        // diagnostics to let the client know to dump that file entirely.
+                        Report(progress, reports, CreateReport(textDocument, diagnostics: null, resultId: null));
+                    }
+                }
+            }
+        }
+
+        private static void Report(IProgress<TReport[]>? progress, ArrayBuilder<TReport> reports, TReport report)
+        {
+            progress?.Report(new[] { report });
+            reports.Add(report);
+        }
+
+        private bool DiagnosticsAreUnchanged(Dictionary<Document, DiagnosticParams> documentToPreviousResult, Document document)
+        {
+            lock (_gate)
+            {
+                var workspace = document.Project.Solution.Workspace;
+                return documentToPreviousResult.TryGetValue(document, out var previousResult) &&
+                       _documentIdToLastResultId.TryGetValue((workspace, document.Id), out var lastReportedResultId) &&
+                       lastReportedResultId == previousResult.PreviousResultId;
+            }
+        }
+
+        private TReport RecordDiagnosticReport(Document document, VSDiagnostic[] diagnostics)
+        {
+            lock (_gate)
+            {
+                // Keep track of the diagnostics we reported here so that we can short-circuit producing diagnostics for
+                // the same diagnostic set in the future.
+                var resultId = _nextResultId++.ToString();
+                _documentIdToLastResultId[(document.Project.Solution.Workspace, document.Id)] = resultId;
+                return CreateReport(ProtocolConversions.DocumentToTextDocumentIdentifier(document), diagnostics, resultId);
+            }
+        }
+
+        protected void AddPreviousResult(
+            Dictionary<Document, DiagnosticParams> documentToDiagnosticParams, DiagnosticParams previousResult)
+        {
+            if (previousResult.TextDocument != null && previousResult.PreviousResultId != null)
+            {
+                var document = _solutionProvider.GetDocument(previousResult.TextDocument);
+                if (document != null)
+                    documentToDiagnosticParams[document] = previousResult;
+            }
+        }
+        public static VSDiagnostic ConvertDiagnostic(Document document, SourceText text, DiagnosticData diagnosticData)
```

Comment:
```suggestion

        public static VSDiagnostic ConvertDiagnostic(Document document, SourceText text, DiagnosticData diagnosticData)
```

Reply:
i'm not seeing what teh difference is here..

Reply:
Blank line before the method :)

(I may have said out loud "got him!" when I spotted it :P)